### PR TITLE
App extension support v2

### PR DESCRIPTION
**Pull request by aramsargsyan**

### CODE REVIEW

File: MatrixKit/Categories/NSBundle+MatrixKit.m [hunk 1]
```diff
@@ -25,6 +25,11 @@ @implementation NSBundle (MatrixKit)
 + (NSBundle*)mxk_assetsBundle
 {
     NSBundle *bundle = [NSBundle bundleForClass:[MXKViewController class]];
+    if ([[bundle.bundleURL pathExtension] isEqualToString:@"appex"]) {
```
Comment (manuroe) [inline MatrixKit/Categories/NSBundle+MatrixKit.m]: Can you create a method in NSBundle+MatrixKit.h like ```mxk_bundleForClass:``` that will do this trick?
We will probably need it in more and more places.

File: MatrixKit/Models/Account/MXKAccountManager.m [hunk 2]
```diff
@@ -344,7 +354,18 @@ - (BOOL)isAPNSAvailable
...
+    
+    NSData *oldAccountData = [[NSUserDefaults standardUserDefaults] objectForKey:kMXKAccountsKey];
+    if (oldAccountData)
+    {
+        [sharedDefaults setObject:oldAccountData forKey:kMXKAccountsKey];
+        [[NSUserDefaults standardUserDefaults] removeObjectForKey:kMXKAccountsKey];
+    }
+    
+    NSData *accountData = [sharedDefaults objectForKey:kMXKAccountsKey];
+    //TEST
```
Comment (manuroe) [inline MatrixKit/Models/Account/MXKAccountManager.m]: useless comment

File: MatrixKit/Views/MXKTableViewCell/MXKTableViewCell.m [hunk 3]
```diff
@@ -22,6 +22,11 @@ + (UINib *)nib
 {
     // Check whether a nib file is available
     NSBundle *mainBundle = [NSBundle bundleForClass:self.class];
+    if ([[mainBundle.bundleURL pathExtension] isEqualToString:@"appex"]) {
```
Comment (manuroe) [inline MatrixKit/Views/MXKTableViewCell/MXKTableViewCell.m]: use ```[NSBundle mxk_bundleForClass:self.class];```

File: MatrixKit/Categories/NSBundle+MatrixKit.m (hunk 1 above)
Comment (manuroe) [inline MatrixKit/Categories/NSBundle+MatrixKit.m]: Do not forget to put { on a new line.

File: MatrixKit/Models/Account/MXKAccountManager.m [hunk 4]
```diff
@@ -17,6 +17,10 @@
 
 #import "MXKAccountManager.h"
 
+static NSString *const kMXKAccountsKey = @"kMXKAccountsKey";
+
+static NSString *const kMXAppGroupID = @"group.org.matrix";
```
Comment (manuroe) [inline MatrixKit/Models/Account/MXKAccountManager.m]: Would it be possible to move this key in a settings of MXKAppSettings (like MXKAppSettings.userDefaultsSuiteName) and let the app define it?

Ideally,  MXKAppSettings should have a property like ```@property  (nonatomic, readonly) NSUserDefaults *userDefaults;``` that returns ```[[NSUserDefaults alloc] initWithSuiteName:kMXAppGroupID];``` or ```[NSUserDefaults standardUserDefaults]``` if not defined (=nil).

Comment (aramsargsyan) [inline MatrixKit/Models/Account/MXKAccountManager.m]: I aggree on moving it to MXKAppSettings, but about the property.. maybe in some places we are not going to use the shared userdefaults in the app?

File: MatrixKit/Categories/NSBundle+MatrixKit.h [hunk 5]
```diff
@@ -55,4 +55,9 @@
  */
 + (NSString *)mxk_localizedStringForKey:(NSString *)key;
 
+/**
+ An AppExtension-compaitable wrapper for bundleForClass.
```
Comment (manuroe) [inline MatrixKit/Categories/NSBundle+MatrixKit.h]: s/compaitable/compatible

File: MatrixKit/Categories/NSBundle+MatrixKit.m [hunk 6]
```diff
@@ -146,4 +142,14 @@ + (NSString *)mxk_localizedStringForKey:(NSString *)key
     return localizedString;
 }
 
++ (NSBundle *)mxk_bundleForClass:(Class)aClass {
```
Comment (manuroe) [inline MatrixKit/Categories/NSBundle+MatrixKit.m]: { a the beginning of line

File: MatrixKit/Models/Account/MXKAccountManager.m (hunk 4 above)
Comment (manuroe) [inline MatrixKit/Models/Account/MXKAccountManager.m]: I agree. We can introduce ```MXKAppSettings. userDefaults``` but we do not have to use it everywhere.

File: MatrixKit/Models/Account/MXKAccountManager.m [hunk 7]
```diff
@@ -80,17 +83,18 @@ - (void)prepareSessionForActiveAccounts
 
 - (void)saveAccounts
 {
+    NSUserDefaults *userDefaults = [MXKAppSettings standardAppSettings].sharedUserDefaults;
     if (mxAccounts.count)
     {
         NSData *accountData = [NSKeyedArchiver archivedDataWithRootObject:mxAccounts];
         
-        [[NSUserDefaults standardUserDefaults] setObject:accountData forKey:@"accounts"];
+        [userDefaults setObject:accountData forKey:kMXKAccountsKey];
```
Comment (manuroe) [inline MatrixKit/Models/Account/MXKAccountManager.m]: You probably need to call the synchronize method to store data for real.

File: MatrixKit/Models/Account/MXKAccountManager.m [hunk 8]
```diff
@@ -344,7 +353,17 @@ - (BOOL)isAPNSAvailable
...
 - (void)loadAccounts
 {
-    NSData *accountData = [[NSUserDefaults standardUserDefaults] objectForKey:@"accounts"];
+    NSUserDefaults *sharedDefaults = [MXKAppSettings standardAppSettings].sharedUserDefaults;
+    
+    NSData *oldAccountData = [[NSUserDefaults standardUserDefaults] objectForKey:kMXKAccountsKey];
+    if (oldAccountData)
+    {
+        [sharedDefaults setObject:oldAccountData forKey:kMXKAccountsKey];
+        [[NSUserDefaults standardUserDefaults] removeObjectForKey:kMXKAccountsKey];
```
Comment (manuroe) [inline MatrixKit/Models/Account/MXKAccountManager.m]: You probably need to call the synchronize method to store data for real.

File: MatrixKit/Models/Account/MXKAccountManager.m [hunk 9]
```diff
@@ -344,7 +353,17 @@ - (BOOL)isAPNSAvailable
 
 - (void)loadAccounts
 {
-    NSData *accountData = [[NSUserDefaults standardUserDefaults] objectForKey:@"accounts"];
+    NSUserDefaults *sharedDefaults = [MXKAppSettings standardAppSettings].sharedUserDefaults;
+    
+    NSData *oldAccountData = [[NSUserDefaults standardUserDefaults] objectForKey:kMXKAccountsKey];
```
Comment (manuroe) [inline MatrixKit/Models/Account/MXKAccountManager.m]: For code maintenance, please a comment explaining that we are a migration of data.

File: MatrixKit/Models/Account/MXKAccountManager.m [hunk 10]
```diff
@@ -344,7 +353,17 @@ - (BOOL)isAPNSAvailable
...
+    if (oldAccountData)
+    {
+        [sharedDefaults setObject:oldAccountData forKey:kMXKAccountsKey];
```
Comment (manuroe) [inline MatrixKit/Models/Account/MXKAccountManager.m]: Please add a NSLog reporting that we are doing an account data migration.

File: MatrixKit/Models/Account/MXKAccountManager.m [hunk 11]
```diff
@@ -344,7 +353,23 @@ - (BOOL)isAPNSAvailable
...
 {
-    NSData *accountData = [[NSUserDefaults standardUserDefaults] objectForKey:@"accounts"];
+    NSUserDefaults *sharedDefaults = [MXKAppSettings standardAppSettings].sharedUserDefaults;
+    
+    //Migration of accountData from [NSUserDefaults standardUserDefaults] to sharedDefaults (shared between apps and extensions)
+    NSData *oldAccountData = [[NSUserDefaults standardUserDefaults] objectForKey:kMXKAccountsKey];
+    if (oldAccountData)
+    {
+        [sharedDefaults setObject:oldAccountData forKey:kMXKAccountsKey];
+        [[NSUserDefaults standardUserDefaults] removeObjectForKey:kMXKAccountsKey];
```
Comment (manuroe) [inline MatrixKit/Models/Account/MXKAccountManager.m]: Before merging, can we temporary comment ```[[NSUserDefaults standardUserDefaults] removeObjectForKey:kMXKAccountsKey];``` out?
The reason is that the user (or the developer as a user) loses his credentials when coming back to an old app version - which is easy when switching between app store, TF and develop versions.

So, please comment the line (and the corresponding one with the synchronize) with a todo comment like: ```// TODO: Erase old location of accountData when the app goes to the app store```

**Pull request merged**